### ISSUE TITLE
Unzipping in place creates an extra nesting of folders

### ISSUE THREAD

**Issue by hydrocheck** [hydroshare/hydroshare#4152]
Unzipping data.zip creates a folder and subfolder (e.g., `data/data/`) instead of the expected single folder (e.g., `data/`).  This is true both for plain folders and zipped aggregation-ready data (e.g., upload zip, unzip it in place, and an extra folder is created).
![014](https://user-images.githubusercontent.com/43424907/108944270-6d552c80-7628-11eb-85c3-516fdcce43c2.png)
[014.zip](https://github.com/hydroshare/hydroshare/files/6033218/014.zip)

**Comment by hyi** [comment]
When looking into the test failures, I start to question whether we want to keep the current behavior with an extra nesting of folders since when creating a new folder to put unzipped files in, we have code to make sure the newly created folder is unique meaning it will not override any existing folder. If we remove this extra folder and directly unzip files in the same folder, the unzipped files could overwrite existing files in the folder that could happen to have the same name. Given that, I think the current way with an extra nesting of folder sounds like a good way to avoid accidental file overwrite. Need to discuss on Monday's meeting @martinseul @dtarb

**Comment by dtarb** [comment]
@hyi @martinseul Extra nesting of folders is really an inconvenience for users and sometimes results in resource corruption.  I think you should try and avoid it if at all possible.  As we have discussed I think there are better ways to handle name conflicts, like just cancelling or blocking the action and letting the user address the problem via renaming.  To inconvenience 98% of the use by extra nesting for perhaps 2% or less name conflict problems, seems like the wrong choice to me.

**Comment by hyi** [comment]
@dtarb @martinseul @sblack-usu The problem is mainly with looking into zip file to get file/folder listing has performance issues especially with large zip file with large amount of files/folders. It is a matter whether we want to look into zip files twice, once for file listing and once for unzipping which definitely has performance implications for the large zip file.

**Comment by dtarb** [comment]
How long does looking into it for listing only take.  I expect it could be quick.

**Comment by hyi** [comment]
It will depend on the zip file mainly on how many nested folders it will have and how many files in each nested folder. I can do some benchmarking using some big files if needed.

**Comment by hyi** [comment]
@dtarb @martinseul @sblack-usu I just realized after our meeting that the zip file is stored in iRODS and iRODS does not provide ability to list files/folders recursively for a zip file. This means if we want to list files/folders recursively for a zip file, we will have to transfer zip file to Django server in order to do it which is definitely not a good idea due to the potential big size of the zip file. So I think we should go with the plan to keep the extra folder while making move operation async. Alternatively, we just add a popup message box when unzipping asking users whether they will want to override existing files or create a extra folder to avoid overriding existing files, and we do what the user selects.

**Comment by dtarb** [comment]
@hyi does iRODS provide the capability to non recursively just list the top level files in a zip file.  I'd be happy to do just a check on these and block unzipping  (or at the users option authorize unzipping with overwrite) if any one of those conflicted.  If there were no conflicts, then the unzip should just proceed silently.

**Comment by hyi** [comment]
@dtarb Confirmed with Terrell that iRODS does not provide this capability. This is response from Terrell: ```iRODS does not know about things inside a zipfile - there are some codepaths that are aware of some types of files - but pretty sure there are limited/no tests around that.```

**Comment by dtarb** [comment]
@hyi I am not the linux guru but looking a few things up what about unzip -l to list the files.  You could strip the front part off the names (before /) and get a unique set to get top level names or just get all the names and check for a name conflict.  Now I guess the problem is that the files are behind irods so this may require an irods rule, but really it does not seem this should be so hard.

**Comment by hyi** [comment]
@dtarb @martinseul I chatted with Terrell about using iRODS iexecmd to run ```unzip -l``` to look into a zip file beforehand for potential file conflicts, and confirmed it can be done if we put unzip binary under ```/var/lib/irods/msiExecCmd_bin``` directory on the iRODS server. We chatted more on the use case and he advised against looking into zip file beforehand and suggested just go ahead to unzip file into a temporary location and then move unzipped files/folders to destination folder, which is a general use case how people handle unzipping. After giving it some thoughts, I like this idea. In our use case, we can unzip initially into an extra folder as we currently do, then check if there could be file/folder conflicts in the destination folder, if no conflicts, we move unzipped files/folders in the extra folder into the destination folder and delete the empty extra folder after the move operation. If there were conflicts, we notify users the conflicting files/folders and keep the unzipped files/folders in the extra folder as we currently do without moving. This essentially combines unzip and move together for users if there are no conflicts and notify users if there are conflicts since users have to handle conflicts anyway. I think this is the best solution to handle this case. If you agree, after I implemented the two move-related issues, i.e., move aggregation move operations to async and handle potential conflicts when moving aggregations into a different folder, this issue can be handled easily.
